Add GSoC 2024 page (#57)
rizinorg/website: content/gsoc/2024.md

The participant will gain an experience of creating comfortable and efficient user interface with
C++/Qt.

### Benefits for the project
It will make interface and user experience more consistent, on par with Rizin itself, and other
tools.

### Assess requirements for midterm/final evaluation
- 1st term: Add scrollbar to necessary widget, improve theming and syntax highlight
- Final term: Managing widgets layouts, docking; provide action status information

### Mentors
- thestr4ng3r
- xvilka
- Megabeets

### Links/Resources
- [User Experience project for Cutter](https://github.com/orgs/rizinorg/project/4/views/1)
- [User Experience project for Rizin](https://github.com/rizinorg/rizin/projects/1)

## Plugins and Python High Level API (175 hour project)

Our current public API to be used by plugin authors is somewhat limited. We need to improve a lot of things about our Plugins support and take it few steps ahead.
This task is only about improving the public C++ and Python interface of Cutter, specifically its graphical user interface components. For a task about exposing Rizin's API for disassembly, analysis and other purposes, see the Rizin bindings task above.

### Task

>  - Expose everything Cutter can offer for plugins authors. This includes high level API, integration of the plugin management etc.

Reviewer: ```suggestion
 - Expose everything Cutter can offer for plugin authors. This includes high level API, integration of the plugin management etc.
```

### Difficulty
Medium

### Benefits for the participant
The participant will gain the understanding on how modern runtimes provide the heap for various
programs, which will be beneficial for the binary exploitation skills.

### Benefits for the project
It will greatly improve the debugging and reverse engineering experience for complex programs,
also provides the way to design the exploitation techniques with the help of Rizin/Cutter.

### Assess requirements for midterm/final evaluation
- 1st term: Design and implement heap visualization widgets, add Rizin test and fixes
- Final term: Various bugfixes related to the heap inspection support on various platforms and allocators, tests and documentation.

### Mentors
- xvilka
- Megabeets

### Links/Resources
- [Issue #1041](https://github.com/rizinorg/cutter/issues/1041)
- [Heap Viewer plugin for IDA Pro](https://github.com/danigargu/heap-viewer)
- [Heap parsing for MacOS, tmalloc, jmalloc](https://github.com/rizinorg/rizin/issues/208)
- [Dynamic Allocator Detection](https://github.com/rizinorg/rizin/issues/157)
- ["heap"-marked Rizin issues](https://github.com/rizinorg/rizin/labels/heap)

> ## Diffing mode (175 hour project)

Reviewer 2: this is going to be implemented by me. we can remove it

Author: Lets keep this for now, we can remove if/when we are accepted.

Reviewer 2: ok

Moreover, the participant will become familiar with the both symbolic and concrete emulation during the
implementation process.

### Benefits for the project
Migrating most architectures will help to deprecate and remove outdated ESIL and will help improving the analysis precision.
Adding uplifting for new architectures that weren't even supported by ESIL will improve the
analysis to even greater degree.

### Assess requirements for midterm/final evaluation
- 1st term: finish the RZIL uplifting for the chosen architecture with basic instruction-level tests
- Final term: implement all changes in the analysis code, added more complex integration tests with types analysis

### Mentors
- xvilka
- thestr4ng3r

### Links/Resources
- [RZIL-labeled issues](https://github.com/rizinorg/rizin/labels/RZIL)
- [ESIL-labeled issues](https://github.com/rizinorg/rizin/labels/RZIL)
- [ESIL to RZIL conversion tracking issue](https://github.com/rizinorg/rizin/issues/2080)
- [Cutter: IL output and graph visual representation](https://github.com/rizinorg/rizin/issues/1133)

> ## FRIDA Integration

Reviewer 2: there are two `Frida integration`. we should change the name of one of these.

Author: Not sure what you are talking about? It has only one.

Reviewer 2: ```suggestion
```

Reviewer 2: i think this was something that was moved further down and this specific section remained.

   debugger
3. Support for missing architectures that are supported by Rizin statically in the BSD native
   debugger
4. Cover more platforms supported by the debugger with automated tests, with CI whenever it's
   possible
5. Fix the bugs in debuggers, minor refactorings of the code

### Skills
* Good knowledge of the C language
* Some experience in debugging with GDB or LLDB

### Difficulty
Hard

### Benefits for the participant
Participant will understand how debugging works on the low level, and will gain experience with variety of different platforms and operating systems.

### Assess requirements for midterm/final evaluation
- 1st term: `SystemZ, MIPS, HPPA support in Linux native, remote GDB debuggers
- Final term: ARM and SPARC support in *BSD debuggers, VAX support in NetBSD

### Mentors
 - xvilka
 - thestr4ng3r
 - ret2libc

### Links/Resources
- [Debug-labeled issues](https://github.com/rizinorg/rizin/labels/Debug)
- [RzDebug-labaled issues](https://github.com/rizinorg/rizin/labels/RzDebug)
- [New Platform support](https://github.com/rizinorg/ideas/labels/New%20Platform)
- [New Architecture support](https://github.com/rizinorg/ideas/labels/New%20Architecture)

> ## FRIDA integration (175 hour project)

Reviewer 2: @XVilka

### Assess requirements for midterm/final evaluation
- 1st term: Implement core of the FRIDA plugin, allowing local and remote debugging features
- Final term: Add support for extended features like calling functions or scripts within the context

### Mentors
 - xvilka
 - thestr4ng3r
 - wargio

### Links/Resources
- [FRIDA](https://frida.re)
- [FRIDA (GitHub)](https://github.com/frida/frida)
- [r2frida](https://github.com/nowsecure/r2frida)

## Rewriting GPL-only code (175 hour project)

Currently some of the Rizin main features rely on the GPL-only code copied from binutils or GCC.
The goal is to rewrite all this code from GPL-only to LGPL or any other less restrictive license.
It is quite important for better adoption of Rizin as a library in other FOSS and commercial
projects.

## Tasks
1. Rewrite C++ demangler to and remove the GPL code
2. Rewrite some of the mainstream architectures that still rely on binutils without using GPL-only
   code

Good example of such architectures are:
- SPARC (there is already capstone-based RzAsm and RzAnalysis plugin but it's less complete than binutils-based one, capstone update is also required)

> - Xtensa (It's better to implemented/updated in capstone)

Reviewer: ```suggestion
- Xtensa (It's better to implement/update it in capstone)
```

- ARC (Same, better implemented in Capstone)

### Skills
Participant should know C and basics of C++ for understanding the mangling scheme

### Difficulty
Medium

### Benefits for the participant
Participant will understand how C++ type information is stored in the name of the methods and classes.

### Assess requirements for midterm/final evaluation
- 1st term: Basic demangling for C++ is rewritten under less restrictive license.
- Final term: At least one binutils-based architecture is reimplemented with more permissive license.

### Mentors
 - xvilka
 - thestr4ng3r
 - wargio

### Links/Resources
- [rz-libdemangle: rewrite/remove GPL-only code](https://github.com/rizinorg/rz-libdemangle/issues/3)
- [Update binutils code to latest](https://github.com/rizinorg/rizin/issues/814)

## Exploitation capabilities improvements (175 hour project)

> Since modern architectures are now enforcing [W^X](https://en.wikipedia.org/wiki/W%5EX), exploiters are using [ROP](https://en.wikipedia.org/wiki/Return-oriented_programming). (Un)fortunately, building ROP chain by hand can be tedious, this is why some tools can be used to ease this construction: ImmunityDBG has [mona.py](https://www.corelan.be/index.php/2012/12/31/jingle-bofs-jingle-rops-sploiting-all-the-things-with-mona-v2/), there is also [ROPgadget](http://www.shell-storm.org/project/ROPgadget/) and [dropper](https://github.com/rizlik/dropper).There exist even tools that can generate ROP chains automatically, for example [exrop](https://github.com/d4em0n/exrop). It's a shame that despite having [ESIL](https://book.rizin.re/disassembling/esil.html), Rizin doesn't have something similar yet. One of the possible solutions would be to build an external plugin or tool which will reuse power of librz and rz-gg. Moreover it makes sense to think about SROP, COOP and BROP support.

Reviewer: Didn't you mean `RzIL` instead of ESIL?

Author: Good catch!

The `rz-gg` tool while has the ability to create a custom shellcode has the outdated database of the shellcodes, so [updating them](https://github.com/rizinorg/rizin/issues/154) is crucial for the tool to be relevant.

### Task

> 1. Update the shellcodes database, imrove `rz-gg` features and documentation

Reviewer: ```suggestion
1. Update the shellcodes database, improve `rz-gg` features and documentation
```

### Mentors
- xvilka
- ret2libc

### Links/Resources
- [ROPGadget](http://shell-storm.org/project/ROPgadget/)
- [Ropper](https://github.com/sashs/Ropper)
- [Angrop](https://github.com/salls/angrop)
- [ROPC](https://github.com/pakt/ropc)
- [exrop](https://github.com/d4em0n/exrop)
- [roper2](https://github.com/oblivia-simplex/roper2)
- [mona.py](https://www.corelan.be/index.php/2012/12/31/jingle-bofs-jingle-rops-sploiting-all-the-things-with-mona-v2/) from corelan
- [Hunting for ROP Gadgets in Style](https://media.blackhat.com/us-13/US-13-Quynh-OptiROP-Hunting-for-ROP-Gadgets-in-Style-WP.pdf) (2012)
- [dropper](https://github.com/rizlik/dropper) a BARF-based rop chain generator
- [Materials](http://dustri.org/b/files/hacklu2014_r2_exploitation.tar.xz) about the exloitation workshop at Hack.lu 2014
- [Slides](https://github.com/XVilka/hacklu) for the exploitation part of workshop at Hack.lu 2015
- [RzEgg related bugs](https://github.com/rizinorg/rizin/labels/RzEgg)

# Microtasks

When taking any of microtasks please be sure someone isn't already working on them, and let
us know if you are going to work on a particular one.

> ## Cutter

Reviewer 2: ```suggestion
## Cutter UX improvements
```

- [Variables and values popup widgets on mouse hover](https://github.com/rizinorg/cutter/issues/2532)
- [Allow to set RzRun profiles from the GUI during debugging](https://github.com/rizinorg/cutter/issues/3199)
- [Double-click on the type in Disasm and Graph widgets should switch to the Types windows and show the selected type](https://github.com/rizinorg/cutter/issues/2780)
- [Set breakpoint inside X-Refs window](https://github.com/rizinorg/cutter/issues/2658)
- [Unified dialogue to set debug symbols servers](https://github.com/rizinorg/cutter/issues/3202)

See full list at our [User Experience](https://github.com/orgs/rizinorg/projects/4/views/1) project covering all parts of RizinOrg: Rizin, Cutter, RzGhidra, rz-pm.

## File formats

Implementing the support for any new file format counts as a microtask. See [New File-Format](https://github.com/rizinorg/ideas/labels/New%20File%20Format) label for pending issues.

## Disassemblers and assemblers

Implementing the support for any new architecture counts as a microtask. See [New-Architecture](https://github.com/rizinorg/ideas/labels/New%20Architecture) label for pending issues.

> ## Analysis

Reviewer 2: ```suggestion
## Rizin analysis improvements
```

Analysis classes, accessible under the `ac` command, is a relatively new feature of rizin.
They provide a way to both manually and automatically manage and use information about classes in the binary.

### Devirtualize method calls using class vtables [#414](https://github.com/rizinorg/rizin/issues/414)

Consider the following call: `call dword [eax + 0x6c]`
Let's assume eax is the base pointer of a vtable we have saved in class analysis and we want to find out the actual address of the called method.

So there should be a command that takes the offset (in this case 0x6c) and looks up the actual destination.
It should be possible to call this command with a specific class, so it only looks into its vtable, or without a class, so it gives a list of possible destinations for all vtables that are not too small for the offset.

When that is implemented, one could also add a command that does the same thing, but automatically takes the offset from the opcode at the current seek.

### Add classes list to `Vb`

`Vb` already supports browsing bin classes. The same thing should be implemented for classes from
analysis.

> ## Refactoring

Reviewer 2: ```suggestion
## RIzin legacy code refactoring
```
